pkg/genericapiserver/config.go

```diff
@@ -343,48 +352,77 @@ func (c Config) New() (*GenericAPIServer, error) {
 	}
 
 	s.installAPI(&c)
-	s.Handler, s.InsecureHandler = s.buildHandlerChains(&c, http.Handler(s.Mux.BaseMux().(*http.ServeMux)))
+
+	// Wire the handler chain:
+	// - requests from InsecureHandler are marked with skipAuth in the context
+	// - requests first go through UnprotectedContainer
+	// - if no route matches:
+	//   * if skipAuth is found in the context, the request goes directly to the ProtectedContainer
+	//   * if skipAuth is not found in the context, the request goes through authn/authz and then
+	//     to the ProtectedContainer
+	protectedHandler := c.ProtectHandlerFunc(s.ProtectedContainer.ServeMux, &c)
+	s.UnprotectedContainer.ServeMux.HandleFunc("/", func(w http.ResponseWriter, req *http.Request) {
+		ctx, ok := c.RequestContextMapper.Get(req)
+		if !ok {
+			internalError(w, req, errors.New("No context found for request"))
+			return
+		}
+		if skipAuthFrom(ctx) {
+			s.ProtectedContainer.ServeHTTP(w, req)
+		} else {
+			protectedHandler.ServeHTTP(w, req)
+		}
+	})
+	s.Handler = c.GenericDecorateHandlerFunc(s.UnprotectedContainer.ServeMux, &c)
+	s.InsecureHandler = c.GenericDecorateHandlerFunc(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		ctx, ok := c.RequestContextMapper.Get(req)
+		if !ok {
+			internalError(w, req, errors.New("No context found for request"))
+			return
+		}
+		c.RequestContextMapper.Update(req, withSkipAuth(ctx))
+		s.UnprotectedContainer.ServeHTTP(w, req)
+	}), &c)
 
 	return s, nil
 }
 
-func (s *GenericAPIServer) buildHandlerChains(c *Config, handler http.Handler) (secure http.Handler, insecure http.Handler) {
```

Comment:
> The downstream construction is more complex unfortunately because it's not a tree.

I'll look more in depth tomorrow morning.  Perhaps we should revisit that.  The downstream code is extremely difficult to read.